File: visualBasic/Ice/session/SessionI.vb
' **********************************************************************
'
' Copyright (c) 2003-2017 ZeroC, Inc. All rights reserved.
'
' **********************************************************************

Imports Demo
Imports System
Imports System.Collections

Public Class SessionI
    Inherits SessionDisp_

    Public Sub New(ByVal name As String)

        _name = name
        _nextId = 0
        _destroy = False
        _objs = New ArrayList

        Console.Out.Write("The session " + _name)
        Console.Out.WriteLine(" is now created.")
    End Sub

    Public Overloads Overrides Function createHello(ByVal c As Ice.Current) As HelloPrx

        SyncLock Me
            If _destroy Then
                Throw New Ice.ObjectNotExistException
            End If

            Dim hello As HelloPrx = HelloPrxHelper.uncheckedCast(c.adapter.addWithUUID(New HelloI(_name, _nextId)))
            _nextId = _nextId + 1
            _objs.Add(hello)
            Return hello
        End SyncLock
    End Function

    Public Overloads Overrides Function getName(ByVal c As Ice.Current) As String
        SyncLock Me
            If _destroy Then
                Throw New Ice.ObjectNotExistException
            End If

            Return _name

        End SyncLock
    End Function

    Public Overloads Overrides Sub destroy(ByVal c As Ice.Current)
        SyncLock Me
            If _destroy Then
                Throw New Ice.ObjectNotExistException
            End If
            _destroy = True

            Console.Out.Write("The session " + _name)
            Console.Out.WriteLine(" is now destroyed.")
            Try
                c.adapter.remove(c.id)
                For Each p As HelloPrx In _objs
                    c.adapter.remove(p.ice_getIdentity())
                Next
            Catch e As Ice.ObjectAdapterDeactivatedException

                ' This method is called on shutdown of the server, in which
                ' case this exception is expected.
            End Try
            _objs.Clear()
        End SyncLock
    End Sub

    Private _name As String
    Private _nextId As Integer ' The per-session id of the next hello object. This is used for tracing purposes.
    Private _objs As ArrayList  ' List of per-session allocated hello objects.
    Private _destroy As Boolean
End Class

File: java-compat/Ice/session/SessionI.java
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

import Demo.*;

class SessionI extends _SessionDisp
{
    public SessionI(String name)
    {
        _name = name;
        System.out.println("The session " + _name + " is now created.");
    }

    @Override
    synchronized public HelloPrx createHello(Ice.Current c)
    {
        if(_destroy)
        {
            throw new Ice.ObjectNotExistException();
        }
        HelloPrx hello = HelloPrxHelper.uncheckedCast(c.adapter.addWithUUID(new HelloI(_name, _nextId++)));
        _objs.add(hello);
        return hello;
    }

    @Override
    synchronized public String getName(Ice.Current c)
    {
        if(_destroy)
        {
            throw new Ice.ObjectNotExistException();
        }
        return _name;
    }

    @Override
    synchronized public void destroy(Ice.Current c)
    {
        if(_destroy)
        {
            throw new Ice.ObjectNotExistException();
        }

        _destroy = true;
        System.out.println("The session " + _name + " is now destroyed.");
        try
        {
            c.adapter.remove(c.id);
            for(HelloPrx p : _objs)
            {
                c.adapter.remove(p.ice_getIdentity());
            }
        }
        catch(Ice.ObjectAdapterDeactivatedException e)
        {
            // This method is called on shutdown of the server, in
            // which case this exception is expected.
        }
        _objs.clear();
    }

    private String _name;
    private boolean _destroy = false; // true if destroy() was called, false otherwise.
    private int _nextId = 0; // The id of the next hello object. This is used for tracing purposes.
    private java.util.List<HelloPrx> _objs =
        new java.util.LinkedList<>(); // List of per-client allocated Hello objects.
}
